File: 235. Lowest Common Ancestor of a Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 // recursive version
// class Solution {
// public:
//     TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
//         if(p->val < root->val && q->val < root->val) {
//             root = lowestCommonAncestor(root->left, p , q);
//         }
//         else if(p->val > root->val && q->val > root->val) {
//             root = lowestCommonAncestor(root->right, p , q);
//         }
//         return root;
//     }
// };
// iterative version
// class Solution {
// public:
//     TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
//         TreeNode* current = root;
//         while(true) {
//             if(p->val < current->val && q->val < current->val) {
//                 current = current->left;
//             }
//             else if(p->val > current->val && q->val > current->val) {
//                 current = current->right;
//             }
//             else return current;
//         }
//     }
// };

File: 235. Lowest Common Ancestor of a Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // p & q < root, go left
        // p & q > root, go right
        while((root.val - p.val) * (root.val - q.val) > 0) {
            root = (p.val < root.val) ? root.left : root.right;
        }
        return root;
    }
}


// /**
//  * Definition for a binary tree node.
//  * public class TreeNode {
//  *     int val;
//  *     TreeNode left;
//  *     TreeNode right;
//  *     TreeNode(int x) { val = x; }
//  * }
//  */
// class Solution {
//     public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
//         if(root.val > p.val && root.val > q.val)return lowestCommonAncestor(root.left, p, q);
//         else if(root.val < p.val && root.val < q.val)return lowestCommonAncestor(root.right, p, q);
//         else return root;
//     }
// }
